Use Daniel Lemire’s fast alternative to modulo reduction

n8sh opened the pull request:
[Full explanation on Daniel Lemire's blog.](https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/) [Another entry covers bias correction when using this method.](https://lemire.me/blog/2016/06/30/fast-random-shuffling/)

If our `gen()` produces `ulong` and we want to convert this to a `uint` in range `0 .. m`, instead of
```
(cast(uint) gen()) % m
```
we can do:
```
cast(uint) (((gen() >>> 32) * m) >>> 32)
```
which replaces a modulo operation with a multiplication and some shifts.

I put a benchmark for this at [bench/randindex_bench.d](https://github.com/n8sh/mir-random/blob/9fe19f33766cdce7baa00b3b024d5c0d2e5bd29b/bench/randindex_bench.d). The performance gain was minor, but the result persisted when running the benchmark repeatedly.

----
**EDIT:**
As @9il points out below, the original benchmark was flawed due to the compiler being too clever and optimizing for a constant modulus. After working to prevent this, the speed advantage of the new version [increased](https://github.com/n8sh/mir-random/blob/2e90600bd905d71653a0735058b296fbe1061222/bench/randindex_bench.d).

9il commented on bench/randindex_bench.d:
```diff
@@ -0,0 +1,152 @@
+T randIndexV2(T, G)(ref G gen, T m)
+    if (isSaturatedRandomEngine!G && isUnsigned!T
+        && (T.sizeof * 2) <= (EngineReturnType!G).sizeof)
+{
```
please mark both with `pragma(inline, false)`. Current benchmark is not fair because LDC is clever enough to optimise algorithm for `100`

9il commented on bench/randindex_bench.d:
```diff
@@ -0,0 +1,152 @@
+    auto randombits = gen() >>> rshift;
+    auto multiresult = randombits * m;
+    auto leftover = cast(T) multiresult;
+    if (leftover < m)
```
after `pragma(inline, false)`, try to replace it with `_expect(leftover < m, false)`, `_expect` is from  `mir.ndslice.internal`

9il commented on bench/randindex_bench.d:
```diff
@@ -0,0 +1,152 @@
+        //sw.stop;
+        //writefln("randIndexV3: %s * 10 ^^ 8 calls/s", double(count) / sw.peek.msecs / 100_000);
+    }
+    writeln("Meaningless sum: ", s);
```
Please add a test that shows that generated numbers are uniformly distributed.

9il commented on source/mir/random/package.d:
```diff
@@ -317,16 +317,46 @@ Returns:
+        enum bool versionD_X32 = true;//On 64-bit machine but using 32-bit pointers.
+    else
+        enum bool versionD_X32 = false;
+    static if (R.sizeof >= T.sizeof * 2 && (size_t.sizeof >= R.sizeof || versionD_X32))
```
- The original algorithm generates the same amount of bits as the result has. It do not need to generate more. - So, it should be just `gen.rand!T`  instead of `gen() >>> rshift` : `rand` will chose lowest or highest bits depends on the engine.
- We only need to check that `multiresult` has at least twice more bits than R.

The problem for this optimisation is in practice `T` is `size_t`. In other hand `x86` and `x86_64` generates extended multiplication results stored in two registers: 32x32->64 and 64x64->128 (x86_64).

- ubyte and ushort T: `randombits * m;` (the result will be 32 bit because of language semantics)
- uint T needs `randombits * cast(ulong) m;` and compiler will do its best job for all targets.
- ulong T needs to use LLVM IR if no `ucent` or `randombits * cast(ucent) m;` if ucent
- ucent T need to use LLVM IR

LLVM IR
 - https://wiki.dlang.org/LDC_inline_IR
 - https://llvm.org/docs/LangRef.html

9il replied:
- Please add `@safe` attribute for unittests.
- Please addd CTFE unittests. LLVM IR can no be executed compile time, so the old code should be used in CTFE (always for simplicity).

9il replied:
`= void` probably unsafe, but I am not sure

n8sh commented:
>please mark both with pragma(inline, false). Current benchmark is not fair because LDC is clever enough to optimise algorithm for 100

Ah! That may explain some of the surprises.

In the next day or two I will revise the test and try again.

9il commented:
Some CPU cache related nitpick for the benchmark: The following code should be before both tests. Also please print results after all tests.
```d
+        foreach(_; 0 .. min(count / 2, 200_000_000u)) //boost CPU
 +        {
 +            s += gen.randIndexV1(cast(uint)100);
 +            s += gen.randIndexV2(cast(uint)100);
 +        }
```

9il commented:
I just realise that we can force saturated engine to be safe. So please do not care about it in this PR.

9il commented:
https://github.com/libmir/mir-random/issues/42

n8sh commented:
>The problem for this optimisation is in practice T is size_t.

I've now added LDC-only code that uses LLVM IR to generate 64-bit size_t. Also, now the method used to reduce the value has nothing to do with `EngineReturnType!G` but only `T`.

n8sh commented:
>Please add a test that shows that generated numbers are uniformly distributed.

Added.

>Also please print results after all tests.

Added.

@9il I think this is everything.

n8sh commented:
@9il 
>`= void` probably unsafe, but I am not sure

Because it's not a pointer it is `@safe` for it to have a void initializer according to D.

9il commented:
needs rebase

9il commented on source/mir/random/package.d:
```diff
@@ -307,6 +321,46 @@ version(mir_random_test) unittest
+                    %b = zext i64 %1 to i128
+                    %ra = mul i128 %a, %b
+                    %rb = bitcast i128 %ra to <2 x i64>
+                    ret <2 x i64> %rb`, core.simd.ulong2)(a, b);
```
Please do not use SIMD here, they makes code slightly more complicated.
`ulong[2]` can be returned. `(a, b, &c, &d)` where ulong c, d; and return `[a, b]`. It is OK to use 64 bit shift, LLVM is clever enough.

9il replied:
In addition you would not need to care Endians

9il commented on bench/randindex_bench.d:
```diff
@@ -0,0 +1,392 @@
+    version (X86_64)
+        private enum bool probablyCanMultiply128 = true;
+    else
+        private enum bool probablyCanMultiply128 = size_t.sizeof >= ulong.sizeof;
```
If I am not wrong, LLVM can always handle 128 / 256 bit multiplication, for targets where it is not supported it will be simulated.

n8sh replied:
I thought so too, but travis-ci compilation failed for x86 on Linux.

9il replied:
Yep, mir-algorithm uses common mul algo now. But it should be faster then div anyway

9il commented on bench/randindex_bench.d:
```diff
@@ -0,0 +1,392 @@
+
+    {
+        //Uniform distribution check.
+        static struct Counter
```
Good idea!

9il commented on bench/randindex_bench.d:
```diff
@@ -0,0 +1,392 @@
+            uint opCall() { return state++; }
+        }
+        import mir.random.engine;
+        enum uint nbuckets = uint(1u << 16);
```
The problem with this number is that `uint.max  % nbuckets  == 0`. So the while loop is never checked.

9il commented on bench/randindex_bench.d:
```diff
@@ -0,0 +1,392 @@
+    }
+
+    {
+        //Uniform distribution check.
```
We should have it in unittests for all unsigned types.

9il commented on source/mir/random/package.d:
```diff
@@ -332,6 +332,44 @@ T rand(T, G)(ref G gen, sizediff_t boundExp = 0)
+    else
+        private enum bool probablyCanMultiply128 = size_t.sizeof >= ulong.sizeof;
+
+    static if (probablyCanMultiply128 && !is(ucent))
```
Please add 256bit for ucent too. ucent division will be very slow

9il commented on bench/randindex_bench.d:
```diff
@@ -0,0 +1,392 @@
+
+    static if (!is(MaybeR == void))
+    {
+        if (!__ctfe)
```
I will do few changes,  please do not force push!

9il commented:
I will add a extended multiplication to mir-algorithm, this will simplify slightly code here.

9il commented:
Please use http://docs.algorithm.dlang.io/latest/mir_utility.html#.extMul in `v0.7.0-alpha4`  both for runtime and compile time code. Old algorithm with division can be removed.

n8sh commented:
@9il
>// LLVM IR by n8sh

BTW if for some reason you want to put my name on other things you can write "Nathan Sashihara".

9il commented:
Added you to the xorshift authors. Please add yourself with this PR into the package.

n8sh commented:
@9il  When compiled by DMD for Windows with --arch=x86_64, `extMul_X86_64` crashes.

```
private ExtMulResult!ulong extMul_X86_64()(ulong a, ulong b)
{
    // pragma(inline, true);
    asm @safe pure nothrow @nogc
    {
        naked;
        mov RAX, RDI;
        mul RSI;
        ret;
    }
}
```

n8sh commented on bench/randindex_bench.d:
```diff
@@ -152,9 +114,9 @@ T randIndexV2(T, G)(ref G gen, T m)
         alias MaybeR = void;
 
-    static if (!is(MaybeR == void))
+    version (LDC) if (!__ctfe)
```
The issue is _expect

9il commented on bench/randindex_bench.d:
```diff
@@ -0,0 +1,379 @@
+            R randombits = cast(R) gen.rand!T;
+            R multiresult = randombits * m;
+            T leftover = cast(T) multiresult;
+            if (_expect(leftover < m, false))
```
OK, then `if` should have most probable condition:
```
...
if (leftover >= m)
    return cast(T) (multiresult >>> finalshift);
...
while(...)
```

n8sh replied:
Is it really a problem to use _expect in runtime code only? We have a unittest that verifies that CTFE and runtime evaluation produce the same result.

n8sh commented:
>Could you please show the output?

Nothing useful:
```
Running .\mir-random-test-library.exe
Program exited with code -1073741819
```

I had to use a debugger to determine where it was crashing.

n8sh commented:
@9il I've written a tiny program that demonstrates the problem on Windows when compiled with `dmd -m64`:

```
import std.traits;

/++
Return type for $(LREF extMul);
+/
struct ExtMulResult(I)
    if (isIntegral!I)
{
    /// Lower I.sizeof * 8 bits
    I low;
    /// Higher I.sizeof * 8 bits
    I high;

}

version(Windows)
version(D_InlineAsm_X86_64)
private ExtMulResult!ulong extMul_X86_64_Windows()(ulong a, ulong b)
{
    asm @nogc nothrow pure @safe
    {
        naked;
        mov RAX, RCX;
        mul RDX;
        ret;
    }
}

void main(string[] args)
{
    import std.stdio;
    auto x = extMul_X86_64_Windows(123,2);
    //This line "works" although the result is garbage:
    writeln(x);
    //This line crashes:
    writeln("extMul_X86_64_Windows(123,2) = ", extMul_X86_64_Windows(123,2));
}
```

9il commented:
> mov RAX, RCX;

This is not like in the original code, it was `mov RAX, RSI`

9il commented:
> EDIT: actually, running this code on a Mac doesn't produce the desired result either, although it doesn't crash. The asm block in extMul_X86_64 seems to be wrong?

This code is wrong. Original has uses RSI register

n8sh commented:
>This is not like in the original code, it was mov RAX, RSI

>This code is wrong for OSX. Original uses RSI register

My mistake -- I changed it for Windows calling conventions and forgot to change it back before testing on another machine.

9il commented:
Great addition!
